lib/ast/dialogues.ts

```diff
@@ -356,20 +382,85 @@ export class DialogueState extends Input {
         this.policy = policy;
 
         this._current = null;
+        this._next = null;
         this.updateCurrent();
     }
 
     /**
-     * The most recently executed history item.
+     * A pointer to the most recently executed ThingTalk statement, if any.
+     *
+     * This is a pointer into the history, provided for API convenience only.
      */
     get current() : DialogueHistoryItem|null {
         return this._current;
     }
 
-    private updateCurrent() {
+    /**
+     * The definition of the main Thingpedia function associated with the current
+     * statement, if any.
+     */
+    get currentFunction() : FunctionDef|null {
+        return this._current ? this._current.stmt.expression.schema : null;
+    }
+
+    /**
+     * The definition of the main Thingpedia query associated with the current
+     * statement, if any.
+     *
+     * If the current statement is a query, this will be the same as
+     * {@link currentFunction}. Otherwise, it will be the last query before
+     * the action is invoked.
+     */
+    get currentQuery() : FunctionDef|null {
+        if (!this._current)
+            return null;
+        const lastQuery = this._current.stmt.expression.lastQuery;
+        if (!lastQuery)
+            return null;
+        return lastQuery.schema;
+    }
+
+    /**
+     * The results of the current ThingTalk statements, if any.
+     *
+     * This is a convenience API to access the results of the current statement.
+     */
+    get currentResults() : DialogueHistoryResultList|null {
+        return this._current ? this._current.results : null;
+    }
+
+    /**
+     * A pointer to the immediate next unexecuted ThingTalk statement, if any.
+     *
+     * This is a pointer into the history, provided for API convenience only.
+     */
+    get next() : DialogueHistoryItem|null {
+        return this._next;
+    }
+
+    /**
+     * The definition of the main Thingpedia function associated with the next
+     * statement, if any.
+     */
+    get nextFunction() : FunctionDef|null {
+        return this._current ? this._current.stmt.expression.schema : null;
```

Comment:
this should be referring to `this._next`, right?

Reply:
Ugh yes